### DISCUSSION TITLE
how to write unit testing for sequelize

### DISCUSSION THREAD

**Discussion by GopalPatidar54** [sequelize/sequelize#15637] [Help & Questions] [answered]
how to write unit testing for Sequelize without a database connection,
how can I create mock 
```

jest.mock('sequelize', () => {
  const actualSequelize = jest.requireActual('sequelize');
  const mSequelize = {
    authenticate: jest.fn(),
    define: jest.fn()
  };
  return {
    Sequelize: jest.fn(() => mSequelize),
    DataTypes: actualSequelize.DataTypes,
    Op: actualSequelize.Op
  };
});
```
I created it like that but it throw error 
 TypeError: Cannot read properties of undefined (reading 'belongsToMany')
Model.belongsToMany(...)

**Comment by ephys** [comment, marked as answer]
You probably don't want to mock the Sequelize class itself. Mock its instance, as that's what is actually going to interact with the database.

You should not mock the `sequelize` package itself. I recommend mocking the module from your code that exposes the Sequelize instance instead.

You should not try to replicate the entire Sequelize instance yourself. Use [`spyOn().mockImplementation()`](https://stackoverflow.com/questions/50091438/jest-how-to-mock-one-specific-method-of-a-class) to mock specific functions of your instance

I can understand mocking the access to the database but you definitely should not mock stuff like `define`, or you won't get actual models (which is why you're getting an error when calling belongsToMany on your model).

For finders, I wouldn't recommend mocking models themselves, it's too complicated. Consider making your own [DAL](https://en.wikipedia.org/wiki/Data_access_layer) with specialized functions such as `getUsers` and mocking those instead.

And finally, consider whether mocking even makes sense for. It's easy to launch a clean docker test DB, and that lets you test that your queries are correct too.
